### PR TITLE
[master] Increase timeout for package download

### PR DESCRIPTION
**Pull request by jongiddy**
## High-level description

Installing packages through the CLI can occasionally timeout, especially on intermittent connections. The CLI exits with a message like: 
```
Error while fetching [https://172.17.0.2/package/resource?url=https://downloads.mesosphere.io/cli/binaries/linux/x86...61d301f571fe26883b5122d567ac4b79bae7febbd2090c81b6cdda523659eb43/dcos-enterprise-cli]: HTTP 504: "Gateway Time-out"
```

This change increases the timeout in the AdminRouter server, while it is acting as a proxy to Cosmos, to reduce the chance of this occurring.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS-38258](https://jira.mesosphere.com/browse/DCOS-38258) Add a longer timeout for package loading

## Related tickets (optional)

Other tickets related to this change:

  - [DCOS-38213](https://jira.mesosphere.com/browse/DCOS-38213) dcoscli fixture times-out while downloading CLI

## Checklist for all PRs

  - [ ] Added a comprehensible changelog entry to `CHANGES.md` or explain why this is not a user-facing change:
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here: Fixes a problem with intermittent network delays, which is difficult to test systematically. I will add comments on manual testing.
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and review process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using GitHub's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

**Comment by mesosphere-mergebot-dev** [comment]
This repo has @mesosphere-mergebot integration. You can interact with the following commands.
```
@mesosphere-mergebot label [Work In Progress |Ready For Review |Holding |Request For Comment |Ship It] 
@mesosphere-mergebot bump-ee  
@mesosphere-mergebot merge-it  
@mesosphere-mergebot override-status  
```

* PR creators can apply one of  `[Ready For Review |Work In Progress]`. [Owners](https://github.com/dcos/dcos/blob/master/owners.json) can apply any label.

**Comment by jongiddy** [comment]
@mesosphere-mergebot bump-ee

**Comment by jongiddy** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by adamtheturtle** [comment]
@jongiddy Does `dcos config set core.timeout` (not recently tested by me) solve your issue without code changes here?

File: packages/adminrouter/extra/src/includes/server/master.conf [hunk 1]
```diff
@@ -331,6 +331,11 @@ location /package/ {
     include includes/disable-request-response-buffering.conf;
     include includes/http-11.conf;
 
+    # Package installation can require large downloads, increasing the chance
+    # for intermittent network connectivity to cause a problem.  Increase
+    # timeouts to make package download more robust.
```
Comment (jgehrcke) [inline packages/adminrouter/extra/src/includes/server/master.conf]: Let's note here that Cosmos buffers the entire "package file" before passing it on to nginx, and that this is an architectural limitation that might be addressed in the future

Comment (jgehrcke) [reply packages/adminrouter/extra/src/includes/server/master.conf]: Let's then also mention DCOS_OSS-3639 in the comment.

**Comment by jgehrcke** [comment]
> Can we have a test for it? We can take care of writing it together with @ahairape .

This easily increases the effort of making this change by an order of magnitude or more. I would like to add that my opinion is that we do not need to test that the nginx' `proxy_read_timeout` directive works as documented :-).

**Comment by jongiddy** [comment]
As well as JP's concern about effort to add a test, is it possible to manipulate the clock that NGINX uses to handle timeouts? Although a test would be great to check that package download can handle a long response delay, it would be excessive to add 5 minutes to CI runtime to test that NGINX configuration works as documented.

I should have noted that I have tested this manually. At home, I regularly hit this problem (504 timeout) using `master`, which currently downloads the Enterprise CLI plugin many times during integration testing. With this PR, I haven't seen the same problem.

**Comment by vespian** [comment]
> This easily increases the effort of making this change by an order of magnitude or more, no? I would like to add that my opinion is that we do not need to test that the nginx' proxy_read_timeout directive works as documented :-).

I understand that it may seem as we are re-testing Nginx's configuration options, both here and in the case of [1].

I try to look at AR not as a Nginx but as a system with many components working together  - it might have been a custom Go app for all that matters (yes, I have seen some discussions about rewriting AR to Go). The test I suggested is a regression test that makes sure that the very valid bug that @jongiddy was hitting does not appear again. The fact that  `proxy_read_timeout` works as intended does not prevent the following from happening in the future:
* there is a human error and somebody removes this config option or "tunes" it in some way
* the openresty/lua code that we write will not start interfering at some point with this setting

What is more - there is an effort to run AR on windows which does not use Nginx at all. Thanks to the strict testing, we can use the current set of tests as acceptance criteria - if the tests pass, we can be pretty confident that the Windows AR will work in concert with DC/OS.

So far this rigorous testing approach allowed me to avoid major bugs and kept me iterating **really fast** - this was not the case when I was taking over AR maintainership. As I am transitioning away from AR, you as the team need to decide if you want to continue this approach, with all its advantages and disadvantages. I am going to remove my thumbs down review and let you decide :)

As for the time - writing this test would take me 30 minutes, pairing with @ahairape  on that would probably increase it to few hours but this is still acceptable I think. The argument raised by @jongiddy wrt. test duration is very valid, but I do not think we need to wait for whole 300s and/or we can parametrize the wait duration and launch tests with different timeouts, tag the tests, or parametrize the timeout like we do with e.g. cache timings.


CC: @jongiddy @ahairape @jgehrcke @gpaul @adamtheturtle @mhrabovcin 

[1] https://jira.mesosphere.com/browse/DCOS-37833

**Comment by jongiddy** [comment]
@mesosphere-mergebot bump-ee

**Comment by mesosphere-mergebot-dev** [comment]
Your pull request's branch is not based on the most recent version of `master`. Please rebase your changes against this repo's `master` branch.

**Comment by jongiddy** [comment]
@mesosphere-mergebot bump-ee

**Comment by jongiddy** [comment]
I've left this without any testing for now. Proper handling of network interruptions and timeouts can best be considered as part of resilience testing, which should be run separately to the approval CI.

**Pull request merged**